### PR TITLE
Initial work on meta validations

### PR DESCRIPTION
**Pull request by starbelly**
- vendor in verl for semver and requirements parsing, comparisons and matching ("port" of Elixir's Version modules to erlang) 
  - added valid meta fields macro definition
  - added required meta fields macro definition
  - added deprecated meta fields macro definition
  - added is_valid_app/2  to hex_tarball
  - added validate_meta/2 to hex_tarball
  - moved create logic into create_tarball/2 (private)
  - changed create/2 to perform validations via is_valid_app/2 before
    attempting to create a tarball
  - updated exists tests to expect a list as the last element in the
    sub-tuple returned via hex_tarball:create/2

### CODE REVIEW

File: rebar.config [hunk 1]
```diff
@@ -10,6 +10,9 @@
     {report_warnings, true},
     {target_erlang_version, 17}
 ]}.
+
+{deps, [{verl, "0.1.0"}]}.
```
Comment (wojtekmach) [inline rebar.config]: due to bootstrapping issues in Hex, we cannot use Hex packages in Hex itself and for that reason we can't use them in hex_core too.

Basically, if we want to use verl we'd need to vendor it in `src/*` and update the vendor script to rewrite module names.

We'd definitely like to have semver validation and I'd be happy to use your package (it's great it's Apache 2.0 licensed) but we'd need to vendor it. Could we vendor it as `hex_version`? WDYT @ericmj?

FYI: we vendor hex_core in Hex (as `mix_hex_*`, to avoid conflicting module names, see e.g.: https://github.com/hexpm/hex/blob/master/src/mix_hex_core.hrl) using [`vendor.sh`](https://github.com/hexpm/hex_core/blob/master/vendor.sh)

File: src/hex_tarball.erl [hunk 2]
```diff
@@ -13,6 +13,33 @@
     {<<"Makefile">>, <<"make">>},
     {<<"Makefile.win">>, <<"make">>}
 ]).
+
+-define(DEPRECATED_META_FIELDS, [
+            <<"maintainers">>,
+            <<"contributors">>
```
Comment (wojtekmach) [inline src/hex_tarball.erl]: please use 4 whitespace indent (we're not super consistent with that, but that's we use most often)

File: src/hex_tarball.erl [hunk 3]
```diff
@@ -13,6 +13,33 @@
...
+]).
+
+-define(META_FIELDS, [
+                       <<"name">>,
+                       <<"version">>,
+                       <<"files">>,
+                       <<"build_tool">>,
+                       <<"build_tools">>
```
Comment (wojtekmach) [inline src/hex_tarball.erl]: the full list of fields is here: https://github.com/hexpm/specifications/blob/master/package_metadata.md.

There's no `build_tool` field - have you encountered it in the wild?

File: src/hex_tarball.erl [hunk 4]
```diff
@@ -13,6 +13,33 @@
...
+]).
+
+-define(REQUIRED_META_FIELDS, [
+                       <<"name">>,
+                       <<"version">>
```
Comment (wojtekmach) [inline src/hex_tarball.erl]: see above mentioned link for list of required fields, we're missing a few (e.g. licenses). That being said, perhaps it's better to focus on adding the simplest possible validations, make sure the code structure and API is good before making this complete - so I guess it's fine if this list is incomplete at this moment.

File: src/hex_tarball.erl [hunk 5]
```diff
@@ -41,33 +68,45 @@
...
     case(byte_size(Tarball) > ?TARBALL_MAX_SIZE) or (UncompressedSize > ?TARBALL_MAX_UNCOMPRESSED_SIZE) of
-        true ->
-            {error, {tarball, too_big}};
+            true ->
+        {error, {tarball, too_big}};
 
-        false ->
-            {ok, {Tarball, Checksum}}
+    false ->
+        {ok, {Tarball, Checksum}}
```
Comment (wojtekmach) [inline src/hex_tarball.erl]: please fix indents.

File: test/hex_tarball_SUITE.erl [hunk 6]
```diff
@@ -21,7 +21,7 @@ memory_test(_Config) ->
       <<"build_tool">> => <<"rebar3">>
      },
     Contents = [{"src/foo.erl", <<"-module(foo).">>}],
-    {ok, {Tarball, Checksum}} = hex_tarball:create(Metadata, Contents),
+    {ok, {Tarball, Checksum, [{<<"maintainers">>,deprecated}]}} = hex_tarball:create(Metadata, Contents),
```
Comment (wojtekmach) [inline test/hex_tarball_SUITE.erl]: could you test the warning in a separate example? The intention of `memory_test` and `disk_test` was to have minimal valid data and just test the memory and disk aspects.

File: rebar.config (hunk 1 above)
Comment (starbelly) [reply rebar.config]: Aye... I just used it to get going... as stated, it's absolutely fine with me if we vendor it in or use something else. It was not clear to me what everyone wanted to do. 

 Also, very much still a WIP : )

File: src/hex_tarball.erl (hunk 3 above)
Comment (starbelly) [reply src/hex_tarball.erl]: In the tests yes ;)  I think I did a push that removes this field and updates the tests.

File: src/hex_tarball.erl (hunk 4 above)
Comment (starbelly) [reply src/hex_tarball.erl]: See latest push.

File: src/hex_tarball.erl (hunk 5 above)
Comment (starbelly) [reply src/hex_tarball.erl]: Will do.

File: src/hex_tarball.erl (hunk 2 above)
Comment (starbelly) [reply src/hex_tarball.erl]: Will do.

File: test/hex_tarball_SUITE.erl (hunk 6 above)
Comment (starbelly) [reply test/hex_tarball_SUITE.erl]: I haven't even put in tests explicitly for the warnings or errors... I only updated the tests so that they would pass. But can remove the maintainers field from this test.

File: rebar.config (hunk 1 above)
Comment (wojtekmach) [reply rebar.config]: it's fine to leave this as is until we make a decision which semver parsing solution we'll go with.

File: src/hex_tarball.erl (hunk 3 above)
Comment (wojtekmach) [reply src/hex_tarball.erl]: haha, oops!

**Comment by starbelly** [comment]
@wojtekmach Do we want to get all the basic validations in now? The reason I ask is I just worked in validating required fields locally, this breaks half the tests. The tests and the behavior therein will have to be altered. As an example if I validate all required fields then this requires me to add them to the meta data at https://github.com/hexpm/hex_core/pull/52/files#diff-f42ce39c05833302551c2271437584cbR43 ... which is fine. 
 However, this in turn requires me to modify the code and basically how we do mock packages for tests. I believe it will require absolute static mocks that live in tests/*_suite_DATA/ . This is also fine, but I don't want to overwhelm anyone with big sweeping to changes to the tests all in one go. I'm ok with that, but need to be sure you are. I can push what I have so you can see the effects yourself.

**Comment by starbelly** [comment]
Disregard the above message, I thought there was going need to be huge changes to the test, but I was wrong. 

  As it stands now we have one failing tests  (build_tools_test) because there is a disagreement in the code. hex_tarball would try to guess the build tool if not list was provided, I'm assuming we don't want this behavior anymore, but not sure.

File: rebar.config (hunk 1 above)
Comment (ericmj) [reply rebar.config]: We could use it into hex_core but we will also need the full feature set from Elixir's `Version` with version comparison and requirement matching, basically it needs to be a 100% reimplementation of `Version`.

File: src/hex_tarball.erl [hunk 7]
```diff
@@ -439,6 +492,76 @@ gzip_no_header(Uncompressed) ->
...
+          ok
+     end;
+validate_meta(has_deprecated, Metadata) ->
+    Keys = maps:keys(Metadata),
+    Fields = ?DEPRECATED_META_FIELDS,
+    find_for(warning, deprecated, Fields, in, Keys);
+validate_meta(has_unknown, Metadata) ->
+    Keys = maps:keys(Metadata),
+    Fields = ?META_FIELDS ++ ?REQUIRED_META_FIELDS ++ ?DEPRECATED_META_FIELDS,
+    find_for(warning, unknown_field, Fields, not_in, Keys);
```
Comment (ericmj) [inline src/hex_tarball.erl]: This should be an error or filtered before included in the tarball.

File: rebar.config (hunk 1 above)
Comment (wojtekmach) [reply rebar.config]: Yeah, for just validating package version, what verl has is enough. However, eventually we'd also want to validate requirements are in valid format too.

We had conversations about doing registry resolution in hex_core which would indeed need the full feature set of Elixir Version module, but that hasn't been decided yet.

Comment (starbelly) [reply rebar.config]: Yup, the plan was to carry the rest over. Notice it has the lexer bit, I wanted to do a 0.1.0 with just the parsing then figure out if I needed to get the rest in. It of course is a port of Elixir's Version module, passing with flying colors! But yeah, let me get the rest in first.

File: src/hex_tarball.erl (hunk 7 above)
Comment (starbelly) [reply src/hex_tarball.erl]: aye... will filter it out.

Comment (starbelly) [reply src/hex_tarball.erl]: Well hmmm... on one hand, it makes sense to ignore and warn for the sake of UX, otoh I consider it an error.

File: rebar.config (hunk 1 above)
Comment (starbelly) [reply rebar.config]: This is done, 100% (structs and string protocol aside ofc)... Give verl a look over, it should have more coverage than the elixir tests atm, but only slightly, it was well tested in elixir. Let me know how you want to vendor it in.

Comment (wojtekmach) [reply rebar.config]: @starbelly verl looks good to me!

Let's vendor it as src/hex_verl*, please include license at the top of the file if it's required (I honestly don't know).

We'd like to have another module that wraps vendored verl. Let's start with minimal module and we'll add more stuff later (e.g. parse_requirement/1 when we actually have such metadata validation)

```erlang
-module(hex_version).
-export([parse/2]).

-type major() :: non_neg_integer().
-type minor() :: non_neg_integer().
-type patch() :: non_neg_integer().
-type pre() :: [binary() | non_neg_integer()].
-type build() :: binary() | undefined.

-type version() :: #{
    major => major(),
    minor => minor(),
    patch => patch(),
    pre   => pre(),
    build => build()
}.

-spec parse(binary()) -> {ok, version()} | {error, invalid_version}.
```

What do you think?

Comment (starbelly) [reply rebar.config]: I am always a fan of wrapping deps 👍

Comment (starbelly) [reply rebar.config]: This In the PR now. I did not add the license to the top, they are both apache 2.0, as to whether it _should_ be done anyway, *shrug*

**Comment by starbelly** [comment]
Ok, as of now all things are passing. I'm sure there are little things to change here and there, but do we want to get anymore validation work in this PR? Or limit it this and do more validation work in a different PR? I'm in favor of the latter, but down for whatever.

**Comment by wojtekmach** [comment]
> do we want to get anymore validation work in this PR? Or limit it this and do more validation work in a different PR? I'm in favor of the latter, but down for whatever.

Agreed, let's work on merging this in before adding more validations, we can always add them later. I'll review the latest changes soon.

Are you planning to integrate this with hex or rebar3? Would be great to integrate end-to-end early and often, but of course it doesn't have to be all on your plate :) I can help with testing this on Hex side.

**Comment by starbelly** [comment]
@wojtekmach  For sure can do a PR for hex and was gonna do one for rebar3 and/or rebar3_code ... less code in the world 🙏  😄

**Comment by ericmj** [comment]
We use binaries everywhere in this API so we should always prefer that over strings.

**Comment by starbelly** [comment]
@ericmj yes s/strings/binaries... so to be clear, you're saying the caller must be do this not hex_core.

**Comment by ericmj** [comment]
Yeah, the caller should do the conversions if they have erlang strings / elixir charlists. In mix/hex's case there would be no conversion because it is already using binaries. If rebar3 is using erlang strings it would have to convert to binaries.

**Comment by starbelly** [comment]
@ericmj I see I see. It was the call I made to normalize_meta/1 before validating... I'll fix this up.

**Comment by starbelly** [comment]
Ok, so per the discussion in #22 I've taken an initial stab at format_error/2 and format_errors/2. format_error/2 works like format_error/1 except that it takes the metadata as the first argument, and error tuple as the second. I've also provided a format_errors/2 which takes the metadata and the error/warnings map. 

  I thought, maybe we could just put the Metadata in the errors map but then I figured the caller will already have this. It'd be nice to have just format_error/1 and include the metadata map in the error tuple, but that sounds hmmm not so good 😄  

 I like the idea of providing specific vars for a specific error and requiring a specific format_error/1 function clause, but I think like the idea of more generic format clauses and leaving what is interpolated with the error/warning message up to the formatter. 

 Of course, if you guys don't agree we can do something more like what already exists.  Also need feedback on what we would like the messages to say. Right now they interpolate nothing and serve as a basis for the rest of the conversation.

**Comment by starbelly** [comment]
After talking to @ferd about the changes in this PR it seems  making this change to create/1 will break compat in a jagged manor. I'm proposing the following to handle this issue:

 - provide create/3 to provide opt in for automatic validation of metadata `create(Meta, Files, validate)`. create/3 can then be checked for existence in rebar3_hex if it exists and use it, otherwise fallback to create/2 . After a good amount of time we could then deprecate create/2. 
 - provide validate_meta/1 so the caller can optionally validate the metadata prior create/1 of their own accord

Thoughts?

**Comment by ericmj** [comment]
Can you elaborate on what the backward compatibility problems are? Why isn't this just a normal breaking API change?

**Comment by starbelly** [comment]
@ericmj yeah, sorry, should have detailed that. I think the concern is if we update hex_core in rebar3 this will break rebar3_hex. Of course, I'll be doing a PR to rebar3_hex to integrate the changes,  but it would be a problem if someone is locked into an old version of rebar3, hex_core,  or rebar3_hex and one is updated and the others are not.

**Comment by ericmj** [comment]
It's a big problem if we can't make break changes to hex_core while it is pre 1.0. I thought rebar3 separate the dependencies of rebar3 and plugins? The risk of conflicts is why mix/hex does not use dependencies or vendors them and renames modules.

I think this is something rebar3 has to solve because being locked to the API we have today would not be feasible for hex_core.

**Comment by ferd** [comment]
We split the libraries between dependencies and plugins, but it's trickier when the library is between a plugin and core rebar3 itself, since we don't make a distinction there.

I mean _we're_ committed to backwards compat, so maybe it was a mistake to rely on a pre-1.0.0 library in the first place?

The other option might be to fold the rebar3_hex plugin within rebar3 so they never have an upgrade conflict. I don't know what @tsloughter would think about that though.

**Comment by ericmj** [comment]
> I mean we're committed to backwards compat, so maybe it was a mistake to rely on a pre-1.0.0 library in the first place?

Rebar3 can still be backwards compatible even if a dependency makes a breaking change. Or is rebar3 supposed to provide its dependencies to plugins as well?

> The other option might be to fold the rebar3_hex plugin within rebar3 so they never have an upgrade conflict. I don't know what @tsloughter would think about that though.

Another option is to vendor hex_core like mix/hex does.

**Comment by ferd** [comment]
Yeah I think that's fine. It will be an acceptable stopgap until we stabilize the rebar3_hex plugin enough to bring it into the main project.

**Comment by starbelly** [comment]
@wojtekmach @ericmj I've squashed everything... still needs to be reviewed and nit picked again, really I suppose mostly the error messages themselves. Other than that, I believe this is ready.

**Comment by starbelly** [comment]
@wojtekmach @ericmj So, hex_core has been vendored in for a bit now. Do we still want this PR?

**Comment by ericmj** [comment]
Definitely still want it. If it's ready for review then I will review it soon.

**Comment by starbelly** [comment]
No rush, just didn't want to leave clutter around.... OCD 😀

File: src/hex_tarball.erl [hunk 8]
```diff
@@ -162,6 +230,20 @@ format_error({checksum_mismatch, ExpectedChecksum, ActualChecksum}) ->
         "Actual   (base16-encoded): ~s",
         [encode_base16(ExpectedChecksum), encode_base16(ActualChecksum)]).
 
+format_error(_Metadata, {Field, is_required}) -> <<Field/binary, <<" is a required metadata field">>/binary >>;
```
Comment (ericmj) [inline src/hex_tarball.erl]: Can we flip these around to `{is_required, Field}`?

File: src/hex_tarball.erl [hunk 9]
```diff
@@ -439,6 +521,74 @@ gzip_no_header(Uncompressed) ->
         zlib:close(Zstream)
     end.
 
+%%====================================================================
+%% Meta-Validation
+%%====================================================================
+-spec is_valid_meta(metadata()) -> {ok, list()} | {error, map()}.
```
Comment (ericmj) [inline src/hex_tarball.erl]: Warnings and errors shouldn't be exclusive. Can the return value instead be `{list(), list()}`? With `{[], []}` meaning no errors and no warnings.

File: src/hex_verl.erl [hunk 10]
```diff
@@ -0,0 +1,205 @@
+%% Vendored from hex_verl v1.0.0, do not edit manually
```
Comment (ericmj) [inline src/hex_verl.erl]: Note to self to prop-test this against Elixir's `Version`.

Comment (ericmj) [reply src/hex_verl.erl]: Can you include a script for us to vendor this dependency so that it's easily repeatable? You can look at https://github.com/hexpm/hex_core/blob/master/vendor.sh for inspiration.

File: src/hex_tarball.erl (hunk 8 above)
Comment (starbelly) [reply src/hex_tarball.erl]: tagged tuples make sense 👍

File: src/hex_tarball.erl (hunk 9 above)
Comment (starbelly) [reply src/hex_tarball.erl]: That makes sense to me.

File: src/hex_verl.erl (hunk 10 above)
Comment (starbelly) [reply src/hex_verl.erl]: Can do.

Comment (starbelly) [reply src/hex_verl.erl]: Oh, hmmm.... did you want a script in hex_core for this? One exists in the verl repo right now.

File: test/hex_tarball_SUITE.erl [hunk 11]
```diff
@@ -126,7 +171,10 @@ build_tools_test(_Config) ->
 
     {ok, #{tarball := Tarball3}} = hex_tarball:create(Metadata, Contents),
     {ok, #{metadata := Metadata2}} = hex_tarball:unpack(Tarball3, memory),
-    false = maps:is_key(<<"build_tools">>, Metadata2),
+
+    %% This check isn't valid anymore since buildtools is required metadata -- Bryan
```
Comment (starbelly) [inline test/hex_tarball_SUITE.erl]: See comment

File: test/hex_tarball_SUITE.erl [hunk 12]
```diff
@@ -224,6 +315,10 @@ unpack_error_handling_test(_Config) ->
       "contents.tar.gz" => <<"badtar">>,
       "CHECKSUM" => <<"63E0D44ED4F61F5A1636A516A6A26890052CE0BB1B1A6EDC66C30282E2EC1A58">>
     },
+
+    {error, {tarball, {inner_checksum_mismatch, _, _}}} = unpack_files(Files5),
+
+    %% I am not sure how to trigger this with the metadata changes. It may be an obsolete check -- Bryan
```
Comment (starbelly) [inline test/hex_tarball_SUITE.erl]: See comment

File: src/hex_tarball.erl [hunk 13]
```diff
@@ -504,3 +586,64 @@ unhex(D) when $a =< D andalso D =< $f ->
...
+validate_meta(has_unknown, Metadata) ->
+    Keys = maps:keys(Metadata),
+    Fields = ?META_FIELDS ++ ?REQUIRED_META_FIELDS ++ ?DEPRECATED_META_FIELDS,
+    search_and_report_for(warning, unknown_field, Fields, not_in, Keys);
+validate_meta(has_required, Metadata) ->
+     Keys = maps:keys(Metadata),
+     Fields = ?REQUIRED_META_FIELDS,
+     search_and_report_for(error, is_required, Keys, not_in, Fields).
+
+search_and_report_for(Type, Msg, Criteria, Op, List) ->
```
Comment (starbelly) [inline src/hex_tarball.erl]: Don't like the name of these search_* functions, but can think of worse things.

File: src/hex_tarball.erl [hunk 14]
```diff
@@ -44,7 +86,34 @@
...
 create(Metadata, Files) ->
+    Normalized = normalize_metadata(Metadata),
+    case is_valid_meta(Normalized) of
+        {[], Warnings} ->
+             Metadata1 = filter_metadata([unknown, deprecated], Metadata, Warnings),
+            {ok, Res} = create_tarball(Metadata1, Files),
+            Final = maps:put(warnings, Warnings, Res),
+            {ok, Final};
+        Res ->
+            {error, Res}
```
Comment (starbelly) [inline src/hex_tarball.erl]: This returns `{error, Errors, Warnings}`

File: src/hex_tarball.erl [hunk 15]
```diff
@@ -45,6 +87,30 @@
 %% @end
 -spec create(metadata(), files()) -> {ok, {tarball(), checksum()}}.
```
Comment (starbelly) [inline src/hex_tarball.erl]: This spec needs to be updated.

File: src/hex_tarball.erl [hunk 16]
```diff
@@ -43,8 +88,32 @@
 %%        inner_checksum => <<178,12,...>>}}
 %% '''
 %% @end
--spec create(metadata(), files()) -> {ok, {tarball(), checksum()}}.
+-spec create(metadata(), files()) -> {ok, tarball()} | {error, errors(), warnings()}.
```
Comment (starbelly) [inline src/hex_tarball.erl]: Adjusted the spec here and the tarball() type. We need to get dialyzer passing, which is not just due to this, but that should be done in another PR vs making this one bigger.

File: src/hex_verl.erl (hunk 10 above)
Comment (ericmj) [reply src/hex_verl.erl]: Yeah, can we vendor the tests also?

File: test/hex_tarball_SUITE.erl (hunk 11 above)
Comment (ericmj) [reply test/hex_tarball_SUITE.erl]: Sorry, I am not following, shouldn't both be valid?

```
true = maps:is_key(<<"build_tools">>, Metadata2),
[] = maps:get(<<"build_tools">>, Metadata2),
```

Or are you saying the `build_tools` needs to be a non-empty list? If so please update the test :)

File: test/hex_tarball_SUITE.erl (hunk 12 above)
Comment (ericmj) [reply test/hex_tarball_SUITE.erl]: It can only happen for an invalid tarball but the checksum needs to match. Maybe it will work with the following checksum:

```elixir
iex(1)> :crypto.hash(:sha256, "badtar") |> Base.encode16()
"8CFEB690D349787B969F3485DE74D3F18FDB877B69C4D29B0E3B09C9F0E4F503"
```

File: src/hex_verl.erl (hunk 10 above)
Comment (starbelly) [reply src/hex_verl.erl]: Can do.

File: test/hex_tarball_SUITE.erl (hunk 11 above)
Comment (starbelly) [reply test/hex_tarball_SUITE.erl]: WIl re-visit 👍

File: test/hex_tarball_SUITE.erl (hunk 12 above)
Comment (starbelly) [reply test/hex_tarball_SUITE.erl]: Will checkout 👍

File: test/support/hex_http_test.erl [hunk 17]
```diff
@@ -87,7 +87,13 @@ fixture(get, <<?TEST_REPO_URL, "/tarballs/ecto-1.0.0.tar">>, _, _) ->
     },
     Metadata = #{
         <<"name">> => <<"ecto">>,
-        <<"version">> => <<"1.0.0">>
+          <<"app">>  => <<"ecto">>,
```
[the remark this answers is not in the document]
Comment (starbelly) [reply test/support/hex_http_test.erl]: Will adjust.

**Comment by ericmj** [comment]
We should also validate that the package is not empty and that the `files` in the metadata matches the package contents. Alternatively we build `files` automatically from the contents.

/cc https://github.com/hexpm/hex/issues/724

**Comment by ericmj** [comment]
We should also do this symlink validation https://github.com/hexpm/hex/issues/725 :)

File: test/hex_tarball_SUITE.erl (hunk 12 above)
Comment (starbelly) [reply test/hex_tarball_SUITE.erl]: aye, that works.

**Comment by starbelly** [comment]
@ericmj symlink validation makes sense to me, will look at this shortly.

File: test/hex_tarball_SUITE.erl (hunk 11 above)
Comment (starbelly) [reply test/hex_tarball_SUITE.erl]: Ahh I see what I was noting here. The test was 

```
 false = maps:is_key(<<"build_tools">>, Metadata2)
```

The test had to change to `true = ...`  and `[] = ...` because build_tools is a required key now. Even if we can't guess it, the key has to be present or validation will fail.

**Comment by starbelly** [comment]
@ericmj Pushed up the first pass on file validations. Some failing tests per enoent errors (i.e., the tests need to be updated to use actual files). However,  I want the mechanics reviewed to be sure we're on the same page. 

- `create/2` now calls `validate/2` which in turn validates files and validates meta and returns a map of `#{files => [..], errors => [...], warnings => [...]}` 
-  valid_files/1 validates that a file exist etc and that if a file is a symlink it must be a safe symlink (this is still in question whether hex allows `../` in symlinks). 
- Files accumulated in valid_files/1 take the form of `{{FileName, FileInfo}, Content}` - the idea here is that this can be passed to create_memory_tarball/1 so it doesn't have to double up on some of the calls. This structure needs revision post-review. 
- Errors and warnings from both valid_files/1 and valid_meta/1 are concatenated within the returned map so the semantics within create stay the same (i.e., if there are errors bail, if their are warnings continue)
- We accumulate the package size along the way by summing content size of a file with the previous accumulated content size, if the content_size ends up being 0, this is an error.

File: src/hex_tarball.erl [hunk 18]
```diff
@@ -523,3 +608,123 @@ unhex(D) when $a =< D andalso D =< $f ->
...
+    case FileInfo of
+        {_, _, symlink, _, _} ->
+            is_safe(FileName, FileInfo);
+        {error, Reason} ->
+            {Reason, FileName};
+        _ ->
+            {ok, {FileName, FileInfo}}
+    end.
+
+is_safe(FileName, FileInfo) ->
```
Comment (starbelly) [inline src/hex_tarball.erl]: Bad function name.

File: src/hex_tarball.erl [hunk 19]
```diff
@@ -610,22 +609,81 @@ unhex(D) when $a =< D andalso D =< $f ->
...
+    {ok, Target} = file:read_link_all(FileName),
+    case filename:safe_relative_path(Target) of
```
Comment (ericmj) [inline src/hex_tarball.erl]: Didn't know about this function. Really great!

**Comment by ericmj** [comment]
> Pushed up the first pass on file validations. Some failing tests per enoent errors (i.e., the tests need to be updated to use actual files). However, I want the mechanics reviewed to be sure we're on the same page.

Everything you listed sounds good 👍

**Comment by starbelly** [comment]
@ericmj Awesome, I'll get the tests updated and code and data structured tidied up 👍

File: src/hex_tarball.erl [hunk 20]
```diff
@@ -43,8 +92,31 @@
...
+-spec create(metadata(), files()) -> {ok, tarball()} | {error, errors(), warnings()}.
 create(Metadata, Files) ->
+    case validate(Metadata, Files) of
+        #{files := _Files1, errors := [], warnings := Warnings} ->
+             Metadata1 = filter_metadata([unknown, deprecated], Metadata, Warnings),
+            {ok, Res} = create_tarball(Metadata1, Files),
+            Final = maps:put(warnings, Warnings, Res),
+            {ok, Final};
+        #{errors := Errors, warnings := Warnings} ->
+            {error, {Errors, Warnings}}
```
Comment (starbelly) [inline src/hex_tarball.erl]: @ericmj I can not remember why we're returning a hash for `ok` and a tuple for `error`, is this cruft or did we want it this way?

**Comment by starbelly** [comment]
@ericmj @wojtekmach Ok, so this needs a lot of clean up and some more unhappy path tests around content size and perhaps symlinks, but before that I think a review is in order to be sure the symlink validation and content size check is proper. 

 The current build is failing for older versions of erlang as they do not have `filename:safe_relative_path/1`.  I'm thinking we can just check the path type as relative_to_cwd/1 is doing most of the work really. `filename:safe_relative_path/1` really just looks at the string vs comparing it to the cwd. Though, it is a nice sanity check to have regardless. If we favor it, we could just put the implementation in the module. Thoughts?

File: src/hex_tarball.erl (hunk 16 above)
Comment (starbelly) [reply src/hex_tarball.erl]: I have adjusted this to function to return a map in both cases.... are we good with that? If not I can revert back to returning a tuple.

File: src/hex_tarball.erl [hunk 21]
```diff
@@ -43,19 +99,42 @@
...
     ContentsTarballCompressed = gzip(ContentsTarball),
     InnerChecksum = inner_checksum(?VERSION, MetadataBinary, ContentsTarballCompressed),
     InnerChecksumBase16 = encode_base16(InnerChecksum),
 
     OuterFiles = [
-       {"VERSION", ?VERSION},
-       {"CHECKSUM", InnerChecksumBase16},
-       {"metadata.config", MetadataBinary},
-       {"contents.tar.gz", ContentsTarballCompressed}
+        #{name => "VERSION", contents => ?VERSION},
```
Comment (starbelly) [inline src/hex_tarball.erl]: I changed this to go with the other `file()` map that is now being slung around to be consistent. I need to add a helper to create a new `file()` map that has all keys if we are in agreement.

File: src/hex_tarball.erl [hunk 22]
```diff
@@ -184,6 +268,19 @@ format_error({checksum_mismatch, ExpectedChecksum, ActualChecksum}) ->
         "Actual   (base16-encoded): ~s",
         [encode_base16(ExpectedChecksum), encode_base16(ActualChecksum)]).
 
+format_error(_Metadata, {is_required, Field}) -> <<Field/binary, <<" is a required metadata field">>/binary >>;
+format_error(_Metadata, {unknown_field, Field}) ->
+    <<Field/binary, <<" is an unknown metadata field and will be ignored">>/binary >>;
+format_error(_Metadata, {deprecated, Field}) ->
+    <<Field/binary, <<" is a deprecated metadata field and will be ignored">>/binary >>;
+format_error(_Metadata, {invalid, <<"version">>}) ->
```
Comment (starbelly) [inline src/hex_tarball.erl]: Need to add `format_error/2` clause for `unsafe_symlink` and `empty_package`

File: src/hex_tarball.erl (hunk 13 above)
Comment (starbelly) [reply src/hex_tarball.erl]: I think we can simplify this by using `--`

**Comment by starbelly** [comment]
@ericmj pang! 😁

**Comment by wojtekmach** [comment]
@starbelly thank you for getting the ball rolling on this. This PR has become stale so I'm going to close this but if you'd like to give it another shot, that'd be appreciated too. Perhaps a better way to move forward is to adding one validation at a time, either the semver check or ever a simpler one, but not both. And once we have a better structure in place, it should be easy to add more checks in which case it'd be ok to add many at a time.

Thanks again!

**Comment by starbelly** [comment]
@wojtekmach That's fine, of course going back through this I see requests for x, y, z which is why it ended up as many vs just a few. I get your rationale though. Happy to re-visit this when I have time.
